File: apps/apple/UCASSignIn/Core/CourseSchedule.swift
import Foundation

public struct CourseScheduleMeeting: Identifiable, Sendable {
    public let id: String
    public let courses: [Course]
    public let day: Date
    public let week: Int
    public let weekday: Int // Monday = 1
    public let start: Date?
    public let end: Date?
    public let classroom: String?
    public let teachers: [String]
    public var hasValidTime: Bool { start != nil && end != nil }
    public let timeText: String

    public func hasEnded(at now: Date) -> Bool {
        guard hasValidTime, let end else { return false }
        return end <= now
    }
}

public struct CourseScheduleProgress: Equatable, Sendable {
    public let total: Int
    public let ended: Int
    public let unknownTime: Int
    public var fraction: Double { total == 0 ? 0 : Double(ended) / Double(total) }
}

public struct CourseScheduleSummary: Identifiable, Sendable {
    public let id: String
    public let weeks: [Int]
    public let weekday: Int
    public let time: String
    public let classroom: String?
    public var weekText: String { CourseSchedulePresentation.weekText(weeks) }
}

/// Read-only projection of actual arrangements, never attendance records or weekly previews.
public struct CourseSchedulePresentation: Sendable {
    public let meetings: [CourseScheduleMeeting]
    public let summaries: [CourseScheduleSummary]
    public let associationIssue: String?

    /// Count merged arrangements, independently of attendance or the current teaching week.
    public func progress(at now: Date) -> CourseScheduleProgress {
        CourseScheduleProgress(total: meetings.count, ended: meetings.filter { $0.hasEnded(at: now) }.count,
                               unknownTime: meetings.filter { !$0.hasValidTime }.count)
    }

    public init(course: CatalogCourse, semester: SchoolSemester, catalog: [CatalogCourse],
                semesters: [SchoolSemester], courses: [Course]) {
        let directory = catalog.filter { $0.semesterId == semester.id }
        let number = course.number.trimmingCharacters(in: .whitespacesAndNewlines)
        let matches = directory.filter { number.isEmpty ? (!$0.id.isEmpty && $0.id == course.id) : $0.number.trimmingCharacters(in: .whitespacesAndNewlines) == number }
        guard matches.count == 1, let target = matches.first else {
            meetings = []; summaries = []
            associationIssue = "暂时无法可靠关联这门课的排课信息。"
            return
        }
        associationIssue = nil
        let values = courses.filter {
            CourseIdentity.catalogCourse(for: $0, catalog: directory, semesters: semesters)?.id == target.id
        }.sorted { ($0.day, $0.beginTime, $0.id) < ($1.day, $1.beginTime, $1.id) }
        var groups: [[String]: [Course]] = [:]
        for value in values {
            guard let day = CourseTime.normalizeDay(value.day),
                  let date = CourseTime.parse(day: day, time: "00:00"),
                  ScheduleCalendar.weekNumber(on: date, in: semester) != nil else { continue }
            let room = Self.clean(value.classroom ?? "")
            let interval = Self.interval(value)
            let key: [String]
            if let (start, end) = interval, !room.isEmpty {
                key = [day, String(start.timeIntervalSince1970), String(end.timeIntervalSince1970), room]
            } else { key = [day, value.id, value.uuid] }
            groups[key, default: []].append(value)
        }
        meetings = groups.values.map { values in
            let first = values[0]
            let day = CourseTime.parse(day: first.day, time: "00:00")!
            let interval = Self.interval(first)
            let room = Self.clean(first.classroom ?? "")
            return CourseScheduleMeeting(id: values.map { "\($0.day)|\($0.id)|\($0.uuid)" }.sorted().joined(separator: ","),
                courses: values, day: day, week: ScheduleCalendar.weekNumber(on: day, in: semester)!,
                weekday: (CourseTime.calendar.component(.weekday, from: day) + 5) % 7 + 1,
                start: interval?.0, end: interval?.1, classroom: room.isEmpty ? nil : room,
                teachers: Array(Set(values.map { Self.clean($0.teacher) }.filter { !$0.isEmpty })).sorted(),
                timeText: Self.timeText(first, interval: interval))
        }.sorted {
            if $0.day != $1.day { return $0.day < $1.day }
            if $0.start != $1.start { return ($0.start ?? .distantFuture) < ($1.start ?? .distantFuture) }
            return $0.id < $1.id
        }
        var summaryGroups: [[String]: [CourseScheduleMeeting]] = [:]
        for value in meetings {
            let key = [String(value.weekday), value.timeText, value.classroom ?? "", value.hasValidTime ? "valid" : value.id]
            summaryGroups[key, default: []].append(value)
        }
        summaries = summaryGroups.map { key, values in
            let value = values[0]
            return CourseScheduleSummary(id: key.joined(separator: "|"), weeks: Array(Set(values.map(\.week))).sorted(),
                weekday: value.weekday, time: value.timeText + (value.hasValidTime ? "" : "（时间待确认）"), classroom: value.classroom)
        }.sorted { ($0.weekday, $0.time, $0.id) < ($1.weekday, $1.time, $1.id) }
    }

    public static func weekText(_ weeks: [Int]) -> String {
        let weeks = Array(Set(weeks)).sorted()
        var spans: [String] = []
        var index = 0
        while index < weeks.count {
            let first = weeks[index]
            var last = first
            index += 1
            while index < weeks.count && weeks[index] == last + 1 { last = weeks[index]; index += 1 }
            spans.append(first == last ? "\(first)" : "\(first)–\(last)")
        }
        return "第 \(spans.joined(separator: "、")) 周"
    }

    private static func timeText(_ course: Course, interval: (Date, Date)?) -> String {
        guard let interval else {
            return "\(course.beginTime.isEmpty ? "暂未提供" : course.beginTime)–\(course.endTime.isEmpty ? "暂未提供" : course.endTime)"
        }
        func clock(_ date: Date) -> String {
            let parts = CourseTime.calendar.dateComponents([.hour, .minute], from: date)
            return String(format: "%02d:%02d", parts.hour!, parts.minute!)
        }
        return "\(clock(interval.0))–\(clock(interval.1))"
    }

    private static func clean(_ value: String) -> String { value.split(whereSeparator: \.isWhitespace).joined(separator: " ") }
    private static func interval(_ course: Course) -> (Date, Date)? {
        guard let start = course.startDate, let end = course.endDate, start < end,
              CourseTime.dayKey(start) == CourseTime.normalizeDay(course.day),
              CourseTime.dayKey(end) == CourseTime.normalizeDay(course.day) else { return nil }
        return (start, end)
    }
}

File: packages/dotnet/UCASSignIn.Core/Courses/CourseSchedule.cs
using System.Collections.Immutable;
using System.Text.Json;

namespace UCASSignIn.Core;

public sealed record CourseScheduleMeeting(string Id, DateOnly Date, int Week, DateTimeOffset? Start,
    string Time, string Classroom, ImmutableArray<string> Teachers, ImmutableArray<string> SourceIds, string ArrangementKey)
{
    public DateTimeOffset? End { get; init; }
    public string Weekday => CourseSchedule.Weekday(Date.DayOfWeek);
    public string TeacherText => Teachers.Length == 0 ? "教师暂未提供" : string.Join("、", Teachers);
    public string ClassroomText => Classroom.Length == 0 ? "教室暂未提供" : Classroom;
    internal bool HasValidTime => Start is { } start && End is { } end && end > start
        && DateOnly.FromDateTime(start.ToOffset(CourseTime.ShanghaiOffset).DateTime) == Date
        && DateOnly.FromDateTime(end.ToOffset(CourseTime.ShanghaiOffset).DateTime) == Date;
    public bool HasEnded(DateTimeOffset now) => HasValidTime && End <= now;
}
public sealed record CourseScheduleSummary(string Weeks, string Weekday, string Time, string Classroom);
public sealed record CourseScheduleProgress(int Total, int Ended, int UnknownTime)
{
    public double Fraction => Total == 0 ? 0 : (double)Ended / Total;
}
public sealed record CourseSchedulePresentation(ImmutableArray<CourseScheduleMeeting> Meetings,
    ImmutableArray<CourseScheduleSummary> Summaries, string? UnavailableReason)
{
    public CourseScheduleProgress ProgressAt(DateTimeOffset now) => new(Meetings.Length,
        Meetings.Count(m => m.HasEnded(now)), Meetings.Count(m => !m.HasValidTime));
}

/// <summary>Only actual schedule records participate; attendance and catalog room descriptions are never inputs.</summary>
public static class CourseSchedule
{
    public static string Weekday(DayOfWeek day) => "周" + "日一二三四五六"[(int)day];
    static string Key(params string?[] fields) => JsonSerializer.Serialize(fields);
    public static bool SameCourse(CatalogCourse a, CatalogCourse b) => a.SemesterId == b.SemesterId
        && a.Id == b.Id && CourseIdentity.Normalize(a.Number) == CourseIdentity.Normalize(b.Number);

    public static CourseSchedulePresentation Build(CatalogCourse target, IEnumerable<SchoolSemester> semesters,
        IEnumerable<CatalogCourse> catalog, IEnumerable<Course> courses)
    {
        var terms = semesters.ToArray();
        var candidates = terms.Where(t => t.Id == target.SemesterId).ToArray();
        if (candidates.Length != 1 || ScheduleCalendar.Range(candidates[0]) is not { } range)
            return new([], [], "无法确定课程所属学期的有效日期范围");
        var term = candidates[0];
        var directory = catalog.ToArray();
        var matching = directory.Where(c => c.SemesterId == target.SemesterId &&
            (CourseIdentity.Normalize(target.Number).Length > 0 ? CourseIdentity.Normalize(c.Number) == CourseIdentity.Normalize(target.Number)
                : !string.IsNullOrWhiteSpace(target.Id) && c.Id == target.Id)).Distinct().ToArray();
        if (matching.Length != 1 || !SameCourse(matching[0], target))
            return new([], [], "课程身份尚未唯一关联，暂时无法确定排课");
        if (terms.Any(t => t.Id != term.Id && ScheduleCalendar.Range(t) is { } other && other.Begin <= range.End && other.End >= range.Begin))
            return new([], [], "学期日期范围重叠，无法可靠关联排课");

        var rows = new List<CourseScheduleMeeting>();
        foreach (var c in courses)
        {
            if (CourseTime.NormalizeDay(c.Day) is not { } day || CourseIdentity.Resolve(c, terms, directory) is not { } resolved
                || !SameCourse(resolved, target)) continue;
            var date = CourseTime.Date(day);
            var start = c.Start?.ToOffset(CourseTime.ShanghaiOffset); var end = c.End?.ToOffset(CourseTime.ShanghaiOffset);
            var valid = start is { } s && end is { } e && e > s && DateOnly.FromDateTime(s.DateTime) == date && DateOnly.FromDateTime(e.DateTime) == date;
            var rawStart = CourseIdentity.Normalize(c.BeginTime); var rawEnd = CourseIdentity.Normalize(c.EndTime);
            var time = valid ? $"{start:HH:mm}–{end:HH:mm}" : rawStart.Length == 0 && rawEnd.Length == 0 ? "暂未提供"
                : $"{(rawStart.Length == 0 ? "暂未提供" : rawStart)}–{(rawEnd.Length == 0 ? "暂未提供" : rawEnd)} · 时间待确认";
            var room = CourseIdentity.Normalize(c.Classroom); var teacher = CourseIdentity.Normalize(c.Teacher);
            var source = Key(day, c.Id, c.Uuid, c.CourseId, c.TeacherId);
            var arrangement = valid ? Key("valid", start!.Value.ToString("HH:mm:ss"), end!.Value.ToString("HH:mm:ss"), room)
                : Key("invalid", rawStart, rawEnd, room);
            // Incomplete records retain their own identity and are never merged with another teacher's record.
            var id = Key(day, arrangement, valid && room.Length > 0 ? null : Key(source, teacher));
            rows.Add(new(id, date, ScheduleCalendar.WeekNumber(date, term)!.Value, valid ? start : null,
                time, room, teacher.Length == 0 ? [] : [teacher], [source], arrangement) { End = valid ? end : null });
        }
        var meetings = rows.GroupBy(r => r.Id).Select(g =>
        {
            var row = g.OrderBy(x => x.SourceIds[0], StringComparer.Ordinal).First();
            var sources = g.SelectMany(x => x.SourceIds).Distinct().Order(StringComparer.Ordinal).ToImmutableArray();
            return row with { Id = Key(row.Date.ToString("yyyyMMdd"), Key(sources.ToArray())), SourceIds = sources,
                Teachers = g.SelectMany(x => x.Teachers).Distinct().Order(StringComparer.Ordinal).ToImmutableArray() };
        }).OrderBy(r => r.Date).ThenBy(r => r.Start is null).ThenBy(r => r.Start).ThenBy(r => r.ArrangementKey, StringComparer.Ordinal)
            .ThenBy(r => r.Id, StringComparer.Ordinal).ToImmutableArray();
        var summaries = meetings.GroupBy(r => (r.Date.DayOfWeek, r.ArrangementKey)).OrderBy(g => ((int)g.Key.DayOfWeek + 6) % 7)
            .ThenBy(g => g.First().Start is null).ThenBy(g => g.First().Start?.TimeOfDay).ThenBy(g => g.Key.ArrangementKey, StringComparer.Ordinal)
            .Select(g => new CourseScheduleSummary(CompressWeeks(g.Select(r => r.Week)), Weekday(g.Key.DayOfWeek), g.First().Time, g.First().ClassroomText)).ToImmutableArray();
        return new(meetings, summaries, null);
    }

    public static string CompressWeeks(IEnumerable<int> source)
    {
        var weeks = source.Distinct().Order().ToArray(); var parts = new List<string>();
        for (var i = 0; i < weeks.Length; i++)
        {
            var first = weeks[i]; var last = first;
            while (i + 1 < weeks.Length && weeks[i + 1] == last + 1) last = weeks[++i];
            parts.Add(first == last ? first.ToString() : $"{first}–{last}");
        }
        return "第 " + string.Join("、", parts) + " 周";
    }
}
